File: 20q-kakao-bot/src/test/kotlin/party/qwer/twentyq/bridge/handlers/UsageHandlerTest.kt
package party.qwer.twentyq.bridge.handlers

import io.mockk.coEvery
import io.mockk.every
import io.mockk.mockk
import kotlinx.coroutines.test.runTest
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Nested
import org.junit.jupiter.api.Test
import party.qwer.twentyq.config.AppProperties
import party.qwer.twentyq.config.properties.Admin
import party.qwer.twentyq.config.properties.GeminiModel
import party.qwer.twentyq.config.properties.Pricing
import party.qwer.twentyq.model.UsagePeriod
import party.qwer.twentyq.rest.LlmRestClient
import party.qwer.twentyq.rest.dto.DailyUsageResponse
import party.qwer.twentyq.rest.dto.UsageListResponse
import party.qwer.twentyq.rest.dto.UsageResponse
import party.qwer.twentyq.service.ExchangeRateService
import party.qwer.twentyq.util.game.GameMessageProvider

/**
 * UsageHandler 테스트
 */
class UsageHandlerTest {
    private val llmRestClient = mockk<LlmRestClient>()
    private val appProperties = mockk<AppProperties>()
    private val messageProvider = mockk<GameMessageProvider>(relaxed = true)
    private val exchangeRateService = mockk<ExchangeRateService>()

    private val handler =
        UsageHandler(
            llmRestClient,
            appProperties,
            messageProvider,
            exchangeRateService,
        )

    init {
        every { appProperties.admin } returns
            mockk<Admin>().apply {
                every { userIds } returns listOf("admin123")
            }
        every { appProperties.pricing } returns
            mockk<Pricing>().apply {
                every { getGeminiModel() } returns GeminiModel.FLASH_25
            }
        coEvery { exchangeRateService.usdToKrw(any()) } answers { firstArg<Double>() * 1400.0 }
        coEvery { exchangeRateService.getRateInfo() } returns "1 USD = 1,400 KRW"
    }

    @Nested
    inner class TodayReport {
        @Test
        fun `should return fetch failed message when daily usage is null`() =
            runTest {
                // Given
                coEvery { llmRestClient.getDailyUsage() } returns null
                every { messageProvider.get("usage.fetch_failed") } returns "사용량 조회 실패"

                // When
                val result = handler.handle("chat123", "admin123", UsagePeriod.TODAY)

                // Then
                assertThat(result).isEqualTo("사용량 조회 실패")
            }

        @Test
        fun `should build daily section with messageProvider when usage exists`() =
            runTest {
                // Given
                val usage =
                    DailyUsageResponse(
                        usageDate = "2025-12-07",
                        inputTokens = 1000L,
                        outputTokens = 500L,
                        reasoningTokens = 100L,
                        totalTokens = 1600L,
                        requestCount = 10L,
                        model = "flash-25",
                    )
                coEvery { llmRestClient.getDailyUsage() } returns usage
                every { messageProvider.get("stats.period.daily") } returns "오늘"
                every { messageProvider.get(any(), *anyVararg()) } answers {
                    val key = firstArg<String>()
                    when (key) {
                        "usage.header_today" -> "📊 토큰 사용량 (오늘)"
                        "usage.label_date" -> "▸ 날짜: 2025-12-07"
                        "usage.label_input_output" -> "  입력: 1,000 / 출력: 500"
                        "usage.label_reasoning" -> "  추론: 100"
                        "usage.label_total" -> "  총: 1,600"
                        "usage.label_request_count" -> "  요청: 10회"
                        "usage.label_cost_header" -> "▸ 예상 비용 (2.5 Flash 기준)"
                        "usage.label_cost_value" -> "  ~₩2"
                        "usage.label_exchange_rate" -> "  (1 USD = 1,400 KRW)"
                        else -> key
                    }
                }

                // When
                val result = handler.handle("chat123", "admin123", UsagePeriod.TODAY)

                // Then
                assertThat(result).contains("📊 토큰 사용량")
                assertThat(result).contains("날짜")
            }
    }

    @Nested
    inner class WeeklyReport {
        @Test
        fun `should return fetch failed weekly message when usage is null`() =
            runTest {
                // Given
                coEvery { llmRestClient.getRecentUsage(7) } returns null
                every { messageProvider.get("usage.fetch_failed_weekly") } returns "주간 사용량 조회 실패"

                // When
                val result = handler.handle("chat123", "admin123", UsagePeriod.WEEKLY)

                // Then
                assertThat(result).isEqualTo("주간 사용량 조회 실패")
            }
    }

    @Nested
    inner class MonthlyReport {
        @Test
        fun `should return fetch failed monthly message when usage is null`() =
            runTest {
                // Given
                coEvery { llmRestClient.getTotalUsageFromDb(30) } returns null
                every { messageProvider.get("usage.fetch_failed_monthly") } returns "월간 사용량 조회 실패"

                // When
                val result = handler.handle("chat123", "admin123", UsagePeriod.MONTHLY)

                // Then
                assertThat(result).isEqualTo("월간 사용량 조회 실패")
            }
    }

    @Nested
    inner class KrwFormatting {
        @Test
        fun `should format cost in KRW using dynamic exchange rate`() =
            runTest {
                // Given
                val usage =
                    DailyUsageResponse(
                        usageDate = "2025-12-07",
                        inputTokens = 1_000_000L,
                        outputTokens = 100_000L,
                        reasoningTokens = 0L,
                        totalTokens = 1_100_000L,
                        requestCount = 50L,
                        model = "flash-25",
                    )
                coEvery { llmRestClient.getDailyUsage() } returns usage

                var capturedCost: String? = null
                every { messageProvider.get("stats.period.daily") } returns "오늘"
                every { messageProvider.get(eq("usage.label_cost_value"), *anyVararg()) } answers {
                    val args = secondArg<Array<Pair<String, Any>>>()
                    capturedCost = args.find { it.first == "cost" }?.second?.toString()
                    "  ~$capturedCost"
                }
                every { messageProvider.get(neq("usage.label_cost_value"), *anyVararg()) } returns "mock"

                // When
                handler.handle("chat123", "admin123", UsagePeriod.TODAY)

                // Then
                // Flash 2.5: input $0.30/1M, output $2.50/1M
                // 1M input = $0.30, 100K output = $0.25, total = $0.55
                // $0.55 * 1400 = ₩770
                assertThat(capturedCost).startsWith("₩")
                assertThat(capturedCost).contains("770")
            }
    }
}

File: game-bot-go/internal/twentyq/service/usage_handler_test.go
package service

import (
	"context"
	"log/slog"
	"os"
	"strings"
	"testing"

	"google.golang.org/grpc"

	"github.com/park285/llm-kakao-bots/game-bot-go/internal/common/llmrest"
	llmv1 "github.com/park285/llm-kakao-bots/game-bot-go/internal/common/llmrest/pb/llm/v1"
	"github.com/park285/llm-kakao-bots/game-bot-go/internal/common/messageprovider"
	"github.com/park285/llm-kakao-bots/game-bot-go/internal/common/testhelper"
	qmodel "github.com/park285/llm-kakao-bots/game-bot-go/internal/twentyq/model"
)

type usageTestEnv struct {
	handler   *UsageHandler
	llmClient *llmrest.Client
	stopLLM   func()
	mocks     struct {
		daily   *llmrest.DailyUsageResponse
		weekly  *llmrest.UsageListResponse
		monthly *llmrest.UsageResponse
	}
}

type fixedExchangeRateService struct {
	rate float64
}

func (s fixedExchangeRateService) UsdToKrw(ctx context.Context, usdAmount float64) float64 {
	return usdAmount * s.rate
}

func (s fixedExchangeRateService) RateInfo(ctx context.Context) string {
	return "1 USD = 1,400 KRW"
}

func setupUsageTestEnv(t *testing.T) *usageTestEnv {
	env := &usageTestEnv{}
	model := "gemini-3-flash-preview"
	stub := &twentyqLLMGRPCStub{
		getDailyUsage: func() (*llmv1.DailyUsageResponse, error) {
			if env.mocks.daily != nil {
				serverModel := ""
				if env.mocks.daily.Model != nil {
					serverModel = *env.mocks.daily.Model
				}
				return &llmv1.DailyUsageResponse{
					UsageDate:       env.mocks.daily.UsageDate,
					InputTokens:     env.mocks.daily.InputTokens,
					OutputTokens:    env.mocks.daily.OutputTokens,
					TotalTokens:     env.mocks.daily.TotalTokens,
					ReasoningTokens: env.mocks.daily.ReasoningTokens,
					RequestCount:    env.mocks.daily.RequestCount,
					Model:           serverModel,
				}, nil
			}
			return &llmv1.DailyUsageResponse{
				UsageDate:       "2023-01-01",
				InputTokens:     1_000_000,
				OutputTokens:    1_000_000,
				TotalTokens:     2_000_000,
				ReasoningTokens: 0,
				RequestCount:    10,
				Model:           model,
			}, nil
		},
		getRecentUsage: func(_ *llmv1.GetRecentUsageRequest) (*llmv1.UsageListResponse, error) {
			if env.mocks.weekly != nil {
				serverModel := ""
				if env.mocks.weekly.Model != nil {
					serverModel = *env.mocks.weekly.Model
				}
				usages := make([]*llmv1.DailyUsageResponse, 0, len(env.mocks.weekly.Usages))
				for _, item := range env.mocks.weekly.Usages {
					itemModel := ""
					if item.Model != nil {
						itemModel = *item.Model
					}
					usages = append(usages, &llmv1.DailyUsageResponse{
						UsageDate:       item.UsageDate,
						InputTokens:     item.InputTokens,
						OutputTokens:    item.OutputTokens,
						TotalTokens:     item.TotalTokens,
						ReasoningTokens: item.ReasoningTokens,
						RequestCount:    item.RequestCount,
						Model:           itemModel,
					})
				}
				return &llmv1.UsageListResponse{
					Usages:            usages,
					TotalInputTokens:  env.mocks.weekly.TotalInputTokens,
					TotalOutputTokens: env.mocks.weekly.TotalOutputTokens,
					TotalTokens:       env.mocks.weekly.TotalTokens,
					TotalRequestCount: env.mocks.weekly.TotalRequestCount,
					Model:             serverModel,
				}, nil
			}
			return &llmv1.UsageListResponse{
				Usages:            nil,
				TotalInputTokens:  7_000_000,
				TotalOutputTokens: 14_000_000,
				TotalTokens:       21_000_000,
				TotalRequestCount: 0,
				Model:             model,
			}, nil
		},
		getTotalUsage: func(_ *llmv1.GetTotalUsageRequest) (*llmv1.UsageResponse, error) {
			if env.mocks.monthly != nil {
				serverModel := ""
				if env.mocks.monthly.Model != nil {
					serverModel = *env.mocks.monthly.Model
				}
				reasoning := int64(0)
				if env.mocks.monthly.ReasoningTokens != nil {
					reasoning = int64(*env.mocks.monthly.ReasoningTokens)
				}
				return &llmv1.UsageResponse{
					InputTokens:     int64(env.mocks.monthly.InputTokens),
					OutputTokens:    int64(env.mocks.monthly.OutputTokens),
					TotalTokens:     int64(env.mocks.monthly.TotalTokens),
					ReasoningTokens: reasoning,
					Model:           serverModel,
				}, nil
			}
			return &llmv1.UsageResponse{
				InputTokens:     1_000_000,
				OutputTokens:    2_000_000,
				TotalTokens:     3_000_000,
				ReasoningTokens: 0,
				Model:           model,
			}, nil
		},
	}

	baseURL, stop := testhelper.StartTestGRPCServer(t, func(s *grpc.Server) {
		llmv1.RegisterLLMServiceServer(s, stub)
	})
	env.stopLLM = stop

	llmClient, err := llmrest.New(llmrest.Config{BaseURL: baseURL})
	if err != nil {
		t.Fatal(err)
	}
	env.llmClient = llmClient
	t.Cleanup(func() {
		_ = llmClient.Close()
	})

	logger := slog.New(slog.NewTextHandler(os.Stdout, nil))
	msgProvider, _ := messageprovider.NewFromYAML(`
stats:
  period:
    daily: "Daily"
usage:
  fetch_failed: "Fetch Failed"
  fetch_failed_weekly: "Fetch Failed Weekly"
  fetch_failed_monthly: "Fetch Failed Monthly"
  header_today: "Today {label}"
  header_weekly: "Weekly {days}"
  header_monthly: "Monthly {days}"
  label_date: "Date: {date}"
  label_input_output: "IO: {input}/{output}"
  label_reasoning: "Reasoning: {reasoning}"
  label_total: "Total: {total}"
  label_request_count: "Reqs: {count}"
  label_sum: "Sum"
  label_input: "Input: {input}"
  label_output: "Output: {output}"
  label_daily_summary: "Day: {date} {total} {count}"
  label_cost_header: "Cost({model})"
  label_cost_value: "CostValue: {cost}"
  label_exchange_rate: "Rate: {rate}"
error:
  no_permission: "Permission Denied"
`)

	exchangeRate := fixedExchangeRateService{rate: 1400.0}
	env.handler = NewUsageHandler(
		[]string{"admin1"},
		llmClient,
		msgProvider,
		exchangeRate,
		logger,
	)

	return env
}

func (e *usageTestEnv) teardown() {
	if e.llmClient != nil {
		_ = e.llmClient.Close()
	}
	if e.stopLLM != nil {
		e.stopLLM()
	}
}

func TestUsageHandler_Permission(t *testing.T) {
	env := setupUsageTestEnv(t)
	defer env.teardown()

	ctx := context.Background()

	// 1. Non-admin
	resp, err := env.handler.Handle(ctx, "chat1", "user1", qmodel.UsagePeriodToday, nil)
	if err != nil {
		t.Errorf("Handle failed: %v", err)
	}
	if !strings.Contains(resp, "Permission Denied") {
		t.Errorf("expected permission denied, got: %s", resp)
	}
}

func TestUsageHandler_Today(t *testing.T) {
	env := setupUsageTestEnv(t)
	defer env.teardown()

	ctx := context.Background()

	resp, err := env.handler.Handle(ctx, "chat1", "admin1", qmodel.UsagePeriodToday, nil)
	if err != nil {
		t.Errorf("Handle failed: %v", err)
	}
	if !strings.Contains(resp, "Today Daily") {
		t.Errorf("expected today usage header, got: %s", resp)
	}
	if !strings.Contains(resp, "Total: 2,000,000") {
		t.Errorf("expected total 2,000,000, got: %s", resp)
	}
	if !strings.Contains(resp, "Cost(3.0 Flash)") {
		t.Errorf("expected cost model, got: %s", resp)
	}
	if !strings.Contains(resp, "CostValue: ₩4,900") {
		t.Errorf("expected cost value, got: %s", resp)
	}
	if !strings.Contains(resp, "Rate: 1 USD = 1,400 KRW") {
		t.Errorf("expected rate info, got: %s", resp)
	}
}

func TestUsageHandler_Weekly(t *testing.T) {
	env := setupUsageTestEnv(t)
	defer env.teardown()
	ctx := context.Background()

	resp, err := env.handler.Handle(ctx, "chat1", "admin1", qmodel.UsagePeriodWeekly, nil)
	if err != nil {
		t.Errorf("Handle failed: %v", err)
	}
	if !strings.Contains(resp, "Weekly 7") {
		t.Errorf("expected weekly usage header, got: %s", resp)
	}
	if !strings.Contains(resp, "Total: 21,000,000") {
		t.Errorf("expected total 21,000,000, got: %s", resp)
	}
	if !strings.Contains(resp, "Cost(3.0 Flash)") {
		t.Errorf("expected cost model, got: %s", resp)
	}
	if !strings.Contains(resp, "CostValue: ₩63,700") {
		t.Errorf("expected cost value, got: %s", resp)
	}
}

func TestUsageHandler_Monthly(t *testing.T) {
	env := setupUsageTestEnv(t)
	defer env.teardown()
	ctx := context.Background()

	resp, err := env.handler.Handle(ctx, "chat1", "admin1", qmodel.UsagePeriodMonthly, nil)
	if err != nil {
		t.Errorf("Handle failed: %v", err)
	}
	if !strings.Contains(resp, "Monthly 30") {
		t.Errorf("expected monthly usage header, got: %s", resp)
	}
	if !strings.Contains(resp, "Total: 3,000,000") {
		t.Errorf("expected total 3,000,000, got: %s", resp)
	}
	if !strings.Contains(resp, "Cost(3.0 Flash)") {
		t.Errorf("expected cost model, got: %s", resp)
	}
	if !strings.Contains(resp, "CostValue: ₩9,100") {
		t.Errorf("expected cost value, got: %s", resp)
	}
}

func TestUsageHandler_ModelOverride(t *testing.T) {
	env := setupUsageTestEnv(t)
	defer env.teardown()
	ctx := context.Background()

	env.mocks.daily = &llmrest.DailyUsageResponse{
		UsageDate:    "2023-01-01",
		InputTokens:  1_000_000,
		OutputTokens: 1_000_000,
		TotalTokens:  2_000_000,
		RequestCount: 10,
	}

	override := "pro-30"
	resp, err := env.handler.Handle(ctx, "chat1", "admin1", qmodel.UsagePeriodToday, &override)
	if err != nil {
		t.Errorf("Handle failed: %v", err)
	}
	if !strings.Contains(resp, "Cost(3.0 Pro)") {
		t.Errorf("expected override model, got: %s", resp)
	}
	if !strings.Contains(resp, "CostValue: ₩19,600") {
		t.Errorf("expected override cost, got: %s", resp)
	}
}

func TestUsageHandler_Errors(t *testing.T) {
	stubErr := &twentyqLLMGRPCStub{hasError: func() bool { return true }}
	baseURL, _ := testhelper.StartTestGRPCServer(t, func(s *grpc.Server) {
		llmv1.RegisterLLMServiceServer(s, stubErr)
	})
	llmClient, err := llmrest.New(llmrest.Config{BaseURL: baseURL})
	if err != nil {
		t.Fatalf("llm client init failed: %v", err)
	}
	t.Cleanup(func() {
		_ = llmClient.Close()
	})
	logger := slog.New(slog.NewTextHandler(os.Stdout, nil))
	msgProvider, _ := messageprovider.NewFromYAML(`
usage:
  fetch_failed: "Fetch Failed"
  fetch_failed_weekly: "Fetch Failed"
  fetch_failed_monthly: "Fetch Failed"
`)

	exchangeRate := fixedExchangeRateService{rate: 1400.0}
	handler := NewUsageHandler([]string{"admin1"}, llmClient, msgProvider, exchangeRate, logger)
	ctx := context.Background()

	t.Run("TodayError", func(t *testing.T) {
		resp, err := handler.Handle(ctx, "c", "admin1", qmodel.UsagePeriodToday, nil)
		if err != nil {
			t.Fatal(err)
		}
		if !strings.Contains(resp, "Fetch Failed") {
			t.Error("expected fetch failed msg")
		}
	})

	t.Run("WeeklyError", func(t *testing.T) {
		resp, err := handler.Handle(ctx, "c", "admin1", qmodel.UsagePeriodWeekly, nil)
		if err != nil {
			t.Fatal(err)
		}
		if !strings.Contains(resp, "Fetch Failed") {
			t.Error("expected fetch failed msg")
		}
	})

	t.Run("MonthlyError", func(t *testing.T) {
		resp, err := handler.Handle(ctx, "c", "admin1", qmodel.UsagePeriodMonthly, nil)
		if err != nil {
			t.Fatal(err)
		}
		if !strings.Contains(resp, "Fetch Failed") {
			t.Error("expected fetch failed msg")
		}
	})
}
